crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -689,17 +1014,55 @@ impl MemoryPool {
         Ok(pool)
     }
 
-    fn get(&self, instance_index: usize) -> impl Iterator<Item = *mut u8> {
+    fn get_base(&self, instance_index: usize, memory_index: usize) -> *mut u8 {
         debug_assert!(instance_index < self.max_instances);
+        debug_assert!(memory_index < self.max_memories);
 
-        let base: *mut u8 = unsafe {
-            self.mapping.as_mut_ptr().add(
-                self.initial_memory_offset + instance_index * self.memory_size * self.max_memories,
-            ) as _
-        };
+        let idx = instance_index * self.max_memories + memory_index;
 
-        let size = self.memory_size;
-        (0..self.max_memories).map(move |i| unsafe { base.add(i * size) })
+        match &self.backing {
+            &PoolBacking::Mmap(ref mapping) => {
+                let offset = self.initial_memory_offset + idx * self.memory_size;
+                unsafe { mapping.as_mut_ptr().offset(offset as isize) }
+            }
+            #[cfg(feature = "memfd-allocator")]
+            &PoolBacking::MemFD(ref slots) => {
+                let slot: &Option<MemFDSlot> = unsafe { slots[idx].get().as_ref().unwrap() };
+                if let &Some(ref slot) = slot {
+                    slot.base as *mut u8
+                } else {
+                    std::ptr::null_mut()
+                }
+            }
+        }
+    }
+
+    fn get<'a>(&'a self, instance_index: usize) -> impl Iterator<Item = *mut u8> + 'a {
+        (0..self.max_memories).map(move |i| self.get_base(instance_index, i))
+    }
+
+    /// Get a mutable borrow of the given MemFD slot.
+    ///
+    /// ## Safety
+    ///
+    /// This returns a mutable borrow on the premise that the caller
+    /// has already allocated a given slot, and has not yet shared it
+    /// outside of the early instance setup. This allows us to mutate
+    /// the slot's state without the overhead of fine-grained locking
+    /// (because we are already covered by the correctness of slot
+    /// allocation).
+    #[cfg(feature = "memfd-allocator")]
+    unsafe fn get_memfd_slot<'a>(
+        &'a self,
+        instance_index: usize,
+        memory_index: usize,
+    ) -> &'a mut Option<MemFDSlot> {
```

Comment:
Yup, I agree. FWIW my original thinking was that we already had the same mutable-access-to-slot thing going on implicitly, because the pooling and uffd allocators were making syscalls to mutate individual slots and relying on higher-level instance mutual exclusion for safety. The difference is just that we didn't have an actual Rust object embodying the slot and actions on it previously.

In any case, the "take ownership and return it" pattern is much nicer; I feel better with this additional level of checking/safety. I ended up doing this with a mutex per slot, just for simplicity and to avoid gratuitous unsafe code and manual box<->raw handling; that's probably acceptable overhead but I'm happy to refactor to actual atomic-exchange magic if not.